### ISSUE TITLE
Use Tuple instead of Vector{Any} in the definition of Optimiser

### ISSUE THREAD

**Issue by cossio** [FluxML/Flux.jl#1823]
Why is `Optimiser` defined like this:

https://github.com/FluxML/Flux.jl/blob/7f375aa46fbdece7ae501f920c12003e9fdf2682/src/optimise/optimisers.jl#L553

This seems to gratuitously violate some of the receommendations in the manual:

https://docs.julialang.org/en/v1/manual/performance-tips/#Avoid-fields-with-abstract-type
https://docs.julialang.org/en/v1/manual/performance-tips/#man-performance-abstract-container

Why not use a `Tuple` here instead? Something like this:

```julia
mutable struct Optimiser{T<:Tuple} <: AbstractOptimiser
  os::T
end

Optimiser(o...) = Optimiser{typeof(o)}(o)
```

**Comment by cossio** [comment]
> I believe the idea was that with decays and such, it might be possible to hit longer tuple lengths which can slow down compile time.

How large can it be?
Also, note that we start froma tuple anyway (the args of `Optimiser(o...)`)

**Comment by ToucheSir** [comment]
I don't believe Vector{Any} is an abstract type, no. Given that this is not a hot code path, I think the burden of proof is on demonstrating the tuple path doesn't cause any meaningful compilation time regressions (see Chain for how bad those can get).

**Comment by DhairyaLGandhi** [comment]
Right, remeber that this is the widest type possible and Julia would use a more specific type to specialise on if it can infer it. So in that way, no type information is actually lost here.

**Comment by CarloLucibello** [comment]
`Flux.Optimiser` has been removed — it now warns and points to `OptimiserChain` from Optimisers.jl, which stores its rules in a `Tuple` exactly as proposed here. Closing.
